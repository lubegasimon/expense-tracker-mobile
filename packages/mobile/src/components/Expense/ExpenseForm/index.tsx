import { View, StyleSheet } from "react-native";
import Input from "../../Input/Input";
import DatePicker from "../../DatePicker";

interface ExpenseFieldProps {
  name: string;
  amount: string;
  details: string;
  category: string;
  createdAt: Date | undefined;
  onChangeName: (value: string) => void;
  onChangeAmount: (value: string) => void;
  onChangeDetails: (value: string) => void;
  onChangeCategory: (value: string) => void;
  onChangeCreatedAt: (date: Date | undefined) => void;
  errors: { [key: string]: string };
}

function ExpenseForm({
  name,
  amount,
  details,
  category,
  createdAt,
  onChangeName,
  onChangeAmount,
  onChangeDetails,
  onChangeCategory,
  onChangeCreatedAt,
  errors,
}: ExpenseFieldProps) {
  return (
    <View style={styles.container}>
      <Input
        label="Name"
        value={name}
        onChangeText={onChangeName}
        placeholder={name}
        error={errors.name || errors.nameRequired}
      />
      <Input
        label="Amount"
        value={amount}
        onChangeText={(value) => onChangeAmount(value)}
        placeholder={amount}
        error={errors.amount || errors.amountRequired}
      />
      <Input
        label="Details"
        value={details}
        onChangeText={(value) => onChangeDetails(value)}
        placeholder={details}
        error={errors.details}
      />
      <Input
        label="Category"
        value={category}
        onChangeText={(value) => onChangeCategory(value)}
        placeholder={category}
        error={errors.category}
      />
      <DatePicker
        date={createdAt}
        onChangeDate={onChangeCreatedAt}
        error={errors.date || errors.dateRequired}
      />
    </View>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: "center",
    alignItems: "center",
    padding: 20,
  },
});

export default ExpenseForm;
