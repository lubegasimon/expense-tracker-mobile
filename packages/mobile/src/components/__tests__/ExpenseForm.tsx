import { render, screen } from "@testing-library/react-native";
import ExpenseForm from "../Expense/ExpenseForm";

/* Hack: Supress React Native animation helpers that scedule timers that conflict with jest */
jest.useFakeTimers();

describe("<ExpenseForm/>", () => {
  const mockProps = {
    name: "Internet",
    amount: "10",
    details: "Internet subscription for February",
    category: "Internet and Communication",
    createdAt: undefined,
    onChangeName: jest.fn(),
    onChangeAmount: jest.fn(),
    onChangeDetails: jest.fn(),
    onChangeCategory: jest.fn(),
    onChangeCreatedAt: jest.fn(),
    errors: {},
  };

  it("should render all text inputs correctly", () => {
    const tree = render(<ExpenseForm {...mockProps} />).toJSON();

    expect(tree).toMatchSnapshot();
    expect(screen.getAllByText("Name")).toBeTruthy();
    expect(screen.getAllByText("Amount")).toBeTruthy();
    expect(screen.getAllByText("Details")).toBeTruthy();
    expect(screen.getAllByText("Category")).toBeTruthy();
  });
});
