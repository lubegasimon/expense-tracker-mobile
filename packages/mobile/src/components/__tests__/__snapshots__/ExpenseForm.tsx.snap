// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ExpenseForm/> should render all text inputs correctly 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "flex": 1,
      "justifyContent": "center",
      "padding": 20,
    }
  }
>
  <View
    style={
      [
        {
          "backgroundColor": "rgba(231, 224, 236, 1)",
          "borderTopLeftRadius": 4,
          "borderTopRightRadius": 4,
        },
        {
          "margin": 5,
          "padding": 10,
          "width": "90%",
        },
      ]
    }
  >
    <View
      collapsable={false}
      style={
        {
          "backgroundColor": "rgba(73, 69, 79, 1)",
          "bottom": 0,
          "height": 1,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "transform": [
            {
              "scaleY": 0.5,
            },
          ],
          "zIndex": 1,
        }
      }
      testID="text-input-underline"
    />
    <View
      onLayout={[Function]}
      style={
        [
          {
            "flexGrow": 1,
            "paddingBottom": 0,
            "paddingTop": 0,
          },
          {
            "minHeight": 20,
          },
        ]
      }
    >
      <View
        pointerEvents="none"
        style={
          [
            {
              "bottom": 0,
              "left": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
            },
            {
              "overflow": "hidden",
            },
          ]
        }
      >
        <View
          collapsable={false}
          pointerEvents="none"
          style={
            {
              "bottom": 0,
              "left": 0,
              "opacity": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
              "transform": [
                {
                  "translateX": 4,
                },
              ],
              "width": 750,
              "zIndex": 3,
            }
          }
        >
          <View
            style={
              {
                "width": 97.33333333333333,
              }
            }
          >
            <Text
              collapsable={false}
              maxFontSizeMultiplier={1.5}
              numberOfLines={1}
              onLayout={[Function]}
              onTextLayout={[Function]}
              style={
                {
                  "color": "rgba(103, 80, 164, 1)",
                  "fontFamily": "System",
                  "fontSize": 16,
                  "fontWeight": undefined,
                  "left": 0,
                  "letterSpacing": 0.15,
                  "lineHeight": undefined,
                  "maxWidth": 97.33333333333333,
                  "opacity": 0,
                  "paddingLeft": 16,
                  "paddingRight": 16,
                  "position": "absolute",
                  "textAlign": "left",
                  "top": 10,
                  "transform": [
                    {
                      "translateX": 0,
                    },
                    {
                      "translateY": 8,
                    },
                    {
                      "scale": 0.75,
                    },
                  ],
                  "writingDirection": "ltr",
                }
              }
              testID="text-input-flat-label-active"
            >
              Name
            </Text>
            <Text
              collapsable={false}
              maxFontSizeMultiplier={1.5}
              numberOfLines={1}
              style={
                {
                  "color": "rgba(73, 69, 79, 1)",
                  "fontFamily": "System",
                  "fontSize": 16,
                  "fontWeight": undefined,
                  "left": 0,
                  "letterSpacing": 0.15,
                  "lineHeight": undefined,
                  "maxWidth": 97.33333333333333,
                  "opacity": 0,
                  "paddingLeft": 16,
                  "paddingRight": 16,
                  "position": "absolute",
                  "textAlign": "left",
                  "top": 10,
                  "transform": [
                    {
                      "translateX": 0,
                    },
                    {
                      "translateY": 8,
                    },
                    {
                      "scale": 0.75,
                    },
                  ],
                  "writingDirection": "ltr",
                }
              }
              testID="text-input-flat-label-inactive"
            >
              Name
            </Text>
          </View>
        </View>
      </View>
      <TextInput
        cursorColor="rgba(103, 80, 164, 1)"
        editable={true}
        maxFontSizeMultiplier={1.5}
        multiline={false}
        onBlur={[Function]}
        onChangeText={[Function]}
        onFocus={[Function]}
        placeholder="Internet"
        placeholderTextColor="transparent"
        selectionColor="rgba(103, 80, 164, 1)"
        style={
          [
            {
              "flexGrow": 1,
              "margin": 0,
            },
            {},
            {
              "paddingBottom": 4,
              "paddingTop": 24,
            },
            {
              "color": "rgba(28, 27, 31, 1)",
              "fontFamily": "System",
              "fontSize": 16,
              "fontWeight": undefined,
              "letterSpacing": 0.15,
              "lineHeight": undefined,
              "minWidth": 65,
              "paddingLeft": 16,
              "paddingRight": 16,
              "textAlign": "left",
              "textAlignVertical": "center",
            },
            false,
            [
              {},
            ],
            undefined,
          ]
        }
        testID="text-input-flat"
        underlineColorAndroid="transparent"
        value="Internet"
      />
    </View>
  </View>
  <View
    style={
      [
        {
          "backgroundColor": "rgba(231, 224, 236, 1)",
          "borderTopLeftRadius": 4,
          "borderTopRightRadius": 4,
        },
        {
          "margin": 5,
          "padding": 10,
          "width": "90%",
        },
      ]
    }
  >
    <View
      collapsable={false}
      style={
        {
          "backgroundColor": "rgba(73, 69, 79, 1)",
          "bottom": 0,
          "height": 1,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "transform": [
            {
              "scaleY": 0.5,
            },
          ],
          "zIndex": 1,
        }
      }
      testID="text-input-underline"
    />
    <View
      onLayout={[Function]}
      style={
        [
          {
            "flexGrow": 1,
            "paddingBottom": 0,
            "paddingTop": 0,
          },
          {
            "minHeight": 20,
          },
        ]
      }
    >
      <View
        pointerEvents="none"
        style={
          [
            {
              "bottom": 0,
              "left": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
            },
            {
              "overflow": "hidden",
            },
          ]
        }
      >
        <View
          collapsable={false}
          pointerEvents="none"
          style={
            {
              "bottom": 0,
              "left": 0,
              "opacity": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
              "transform": [
                {
                  "translateX": 4,
                },
              ],
              "width": 750,
              "zIndex": 3,
            }
          }
        >
          <View
            style={
              {
                "width": 97.33333333333333,
              }
            }
          >
            <Text
              collapsable={false}
              maxFontSizeMultiplier={1.5}
              numberOfLines={1}
              onLayout={[Function]}
              onTextLayout={[Function]}
              style={
                {
                  "color": "rgba(103, 80, 164, 1)",
                  "fontFamily": "System",
                  "fontSize": 16,
                  "fontWeight": undefined,
                  "left": 0,
                  "letterSpacing": 0.15,
                  "lineHeight": undefined,
                  "maxWidth": 97.33333333333333,
                  "opacity": 0,
                  "paddingLeft": 16,
                  "paddingRight": 16,
                  "position": "absolute",
                  "textAlign": "left",
                  "top": 10,
                  "transform": [
                    {
                      "translateX": 0,
                    },
                    {
                      "translateY": 8,
                    },
                    {
                      "scale": 0.75,
                    },
                  ],
                  "writingDirection": "ltr",
                }
              }
              testID="text-input-flat-label-active"
            >
              Amount
            </Text>
            <Text
              collapsable={false}
              maxFontSizeMultiplier={1.5}
              numberOfLines={1}
              style={
                {
                  "color": "rgba(73, 69, 79, 1)",
                  "fontFamily": "System",
                  "fontSize": 16,
                  "fontWeight": undefined,
                  "left": 0,
                  "letterSpacing": 0.15,
                  "lineHeight": undefined,
                  "maxWidth": 97.33333333333333,
                  "opacity": 0,
                  "paddingLeft": 16,
                  "paddingRight": 16,
                  "position": "absolute",
                  "textAlign": "left",
                  "top": 10,
                  "transform": [
                    {
                      "translateX": 0,
                    },
                    {
                      "translateY": 8,
                    },
                    {
                      "scale": 0.75,
                    },
                  ],
                  "writingDirection": "ltr",
                }
              }
              testID="text-input-flat-label-inactive"
            >
              Amount
            </Text>
          </View>
        </View>
      </View>
      <TextInput
        cursorColor="rgba(103, 80, 164, 1)"
        editable={true}
        maxFontSizeMultiplier={1.5}
        multiline={false}
        onBlur={[Function]}
        onChangeText={[Function]}
        onFocus={[Function]}
        placeholder="10"
        placeholderTextColor="transparent"
        selectionColor="rgba(103, 80, 164, 1)"
        style={
          [
            {
              "flexGrow": 1,
              "margin": 0,
            },
            {},
            {
              "paddingBottom": 4,
              "paddingTop": 24,
            },
            {
              "color": "rgba(28, 27, 31, 1)",
              "fontFamily": "System",
              "fontSize": 16,
              "fontWeight": undefined,
              "letterSpacing": 0.15,
              "lineHeight": undefined,
              "minWidth": 65,
              "paddingLeft": 16,
              "paddingRight": 16,
              "textAlign": "left",
              "textAlignVertical": "center",
            },
            false,
            [
              {},
            ],
            undefined,
          ]
        }
        testID="text-input-flat"
        underlineColorAndroid="transparent"
        value="10"
      />
    </View>
  </View>
  <View
    style={
      [
        {
          "backgroundColor": "rgba(231, 224, 236, 1)",
          "borderTopLeftRadius": 4,
          "borderTopRightRadius": 4,
        },
        {
          "margin": 5,
          "padding": 10,
          "width": "90%",
        },
      ]
    }
  >
    <View
      collapsable={false}
      style={
        {
          "backgroundColor": "rgba(73, 69, 79, 1)",
          "bottom": 0,
          "height": 1,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "transform": [
            {
              "scaleY": 0.5,
            },
          ],
          "zIndex": 1,
        }
      }
      testID="text-input-underline"
    />
    <View
      onLayout={[Function]}
      style={
        [
          {
            "flexGrow": 1,
            "paddingBottom": 0,
            "paddingTop": 0,
          },
          {
            "minHeight": 20,
          },
        ]
      }
    >
      <View
        pointerEvents="none"
        style={
          [
            {
              "bottom": 0,
              "left": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
            },
            {
              "overflow": "hidden",
            },
          ]
        }
      >
        <View
          collapsable={false}
          pointerEvents="none"
          style={
            {
              "bottom": 0,
              "left": 0,
              "opacity": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
              "transform": [
                {
                  "translateX": 4,
                },
              ],
              "width": 750,
              "zIndex": 3,
            }
          }
        >
          <View
            style={
              {
                "width": 97.33333333333333,
              }
            }
          >
            <Text
              collapsable={false}
              maxFontSizeMultiplier={1.5}
              numberOfLines={1}
              onLayout={[Function]}
              onTextLayout={[Function]}
              style={
                {
                  "color": "rgba(103, 80, 164, 1)",
                  "fontFamily": "System",
                  "fontSize": 16,
                  "fontWeight": undefined,
                  "left": 0,
                  "letterSpacing": 0.15,
                  "lineHeight": undefined,
                  "maxWidth": 97.33333333333333,
                  "opacity": 0,
                  "paddingLeft": 16,
                  "paddingRight": 16,
                  "position": "absolute",
                  "textAlign": "left",
                  "top": 10,
                  "transform": [
                    {
                      "translateX": 0,
                    },
                    {
                      "translateY": 8,
                    },
                    {
                      "scale": 0.75,
                    },
                  ],
                  "writingDirection": "ltr",
                }
              }
              testID="text-input-flat-label-active"
            >
              Details
            </Text>
            <Text
              collapsable={false}
              maxFontSizeMultiplier={1.5}
              numberOfLines={1}
              style={
                {
                  "color": "rgba(73, 69, 79, 1)",
                  "fontFamily": "System",
                  "fontSize": 16,
                  "fontWeight": undefined,
                  "left": 0,
                  "letterSpacing": 0.15,
                  "lineHeight": undefined,
                  "maxWidth": 97.33333333333333,
                  "opacity": 0,
                  "paddingLeft": 16,
                  "paddingRight": 16,
                  "position": "absolute",
                  "textAlign": "left",
                  "top": 10,
                  "transform": [
                    {
                      "translateX": 0,
                    },
                    {
                      "translateY": 8,
                    },
                    {
                      "scale": 0.75,
                    },
                  ],
                  "writingDirection": "ltr",
                }
              }
              testID="text-input-flat-label-inactive"
            >
              Details
            </Text>
          </View>
        </View>
      </View>
      <TextInput
        cursorColor="rgba(103, 80, 164, 1)"
        editable={true}
        maxFontSizeMultiplier={1.5}
        multiline={false}
        onBlur={[Function]}
        onChangeText={[Function]}
        onFocus={[Function]}
        placeholder="Internet subscription for February"
        placeholderTextColor="transparent"
        selectionColor="rgba(103, 80, 164, 1)"
        style={
          [
            {
              "flexGrow": 1,
              "margin": 0,
            },
            {},
            {
              "paddingBottom": 4,
              "paddingTop": 24,
            },
            {
              "color": "rgba(28, 27, 31, 1)",
              "fontFamily": "System",
              "fontSize": 16,
              "fontWeight": undefined,
              "letterSpacing": 0.15,
              "lineHeight": undefined,
              "minWidth": 65,
              "paddingLeft": 16,
              "paddingRight": 16,
              "textAlign": "left",
              "textAlignVertical": "center",
            },
            false,
            [
              {},
            ],
            undefined,
          ]
        }
        testID="text-input-flat"
        underlineColorAndroid="transparent"
        value="Internet subscription for February"
      />
    </View>
  </View>
  <View
    style={
      [
        {
          "backgroundColor": "rgba(231, 224, 236, 1)",
          "borderTopLeftRadius": 4,
          "borderTopRightRadius": 4,
        },
        {
          "margin": 5,
          "padding": 10,
          "width": "90%",
        },
      ]
    }
  >
    <View
      collapsable={false}
      style={
        {
          "backgroundColor": "rgba(73, 69, 79, 1)",
          "bottom": 0,
          "height": 1,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "transform": [
            {
              "scaleY": 0.5,
            },
          ],
          "zIndex": 1,
        }
      }
      testID="text-input-underline"
    />
    <View
      onLayout={[Function]}
      style={
        [
          {
            "flexGrow": 1,
            "paddingBottom": 0,
            "paddingTop": 0,
          },
          {
            "minHeight": 20,
          },
        ]
      }
    >
      <View
        pointerEvents="none"
        style={
          [
            {
              "bottom": 0,
              "left": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
            },
            {
              "overflow": "hidden",
            },
          ]
        }
      >
        <View
          collapsable={false}
          pointerEvents="none"
          style={
            {
              "bottom": 0,
              "left": 0,
              "opacity": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
              "transform": [
                {
                  "translateX": 4,
                },
              ],
              "width": 750,
              "zIndex": 3,
            }
          }
        >
          <View
            style={
              {
                "width": 97.33333333333333,
              }
            }
          >
            <Text
              collapsable={false}
              maxFontSizeMultiplier={1.5}
              numberOfLines={1}
              onLayout={[Function]}
              onTextLayout={[Function]}
              style={
                {
                  "color": "rgba(103, 80, 164, 1)",
                  "fontFamily": "System",
                  "fontSize": 16,
                  "fontWeight": undefined,
                  "left": 0,
                  "letterSpacing": 0.15,
                  "lineHeight": undefined,
                  "maxWidth": 97.33333333333333,
                  "opacity": 0,
                  "paddingLeft": 16,
                  "paddingRight": 16,
                  "position": "absolute",
                  "textAlign": "left",
                  "top": 10,
                  "transform": [
                    {
                      "translateX": 0,
                    },
                    {
                      "translateY": 8,
                    },
                    {
                      "scale": 0.75,
                    },
                  ],
                  "writingDirection": "ltr",
                }
              }
              testID="text-input-flat-label-active"
            >
              Category
            </Text>
            <Text
              collapsable={false}
              maxFontSizeMultiplier={1.5}
              numberOfLines={1}
              style={
                {
                  "color": "rgba(73, 69, 79, 1)",
                  "fontFamily": "System",
                  "fontSize": 16,
                  "fontWeight": undefined,
                  "left": 0,
                  "letterSpacing": 0.15,
                  "lineHeight": undefined,
                  "maxWidth": 97.33333333333333,
                  "opacity": 0,
                  "paddingLeft": 16,
                  "paddingRight": 16,
                  "position": "absolute",
                  "textAlign": "left",
                  "top": 10,
                  "transform": [
                    {
                      "translateX": 0,
                    },
                    {
                      "translateY": 8,
                    },
                    {
                      "scale": 0.75,
                    },
                  ],
                  "writingDirection": "ltr",
                }
              }
              testID="text-input-flat-label-inactive"
            >
              Category
            </Text>
          </View>
        </View>
      </View>
      <TextInput
        cursorColor="rgba(103, 80, 164, 1)"
        editable={true}
        maxFontSizeMultiplier={1.5}
        multiline={false}
        onBlur={[Function]}
        onChangeText={[Function]}
        onFocus={[Function]}
        placeholder="Internet and Communication"
        placeholderTextColor="transparent"
        selectionColor="rgba(103, 80, 164, 1)"
        style={
          [
            {
              "flexGrow": 1,
              "margin": 0,
            },
            {},
            {
              "paddingBottom": 4,
              "paddingTop": 24,
            },
            {
              "color": "rgba(28, 27, 31, 1)",
              "fontFamily": "System",
              "fontSize": 16,
              "fontWeight": undefined,
              "letterSpacing": 0.15,
              "lineHeight": undefined,
              "minWidth": 65,
              "paddingLeft": 16,
              "paddingRight": 16,
              "textAlign": "left",
              "textAlignVertical": "center",
            },
            false,
            [
              {},
            ],
            undefined,
          ]
        }
        testID="text-input-flat"
        underlineColorAndroid="transparent"
        value="Internet and Communication"
      />
    </View>
  </View>
  <RNCSafeAreaProvider
    onInsetsChange={[Function]}
    style={
      [
        {
          "flex": 1,
        },
        undefined,
      ]
    }
  />
</View>
`;
